src/lib.rs

```diff
@@ -416,12 +442,19 @@ impl<T: Translation> Mapping<T> {
     /// and modifying those would violate architectural break-before-make (BBM) requirements.
     pub fn modify_range<F>(&mut self, range: &MemoryRegion, f: &F) -> Result<(), MapError>
     where
-        F: Fn(&MemoryRegion, &mut Descriptor, usize) -> Result<(), ()> + ?Sized,
+        F: Fn(&MemoryRegion, &mut Descriptor, usize) -> Result<bool, ()> + ?Sized,
     {
         if self.active() {
             self.check_range_bbm(range, f)?;
         }
-        self.root.modify_range(range, f)?;
+        if self.root.modify_range(range, f, self.active())? && self.active() {
```

Comment:
If `modify_range` returns an error then the barrier below won't apply. Is that correct? What if the range has already been partially modified before the error occurs?

Reply:
The limbo state that results from an error occurring halfway is architecturally sound, because only BBM-safe updates are permitted when the mapping is active. This means that the resulting state is a mix of the old state and the new state, which may be an error condition according to the semantics of the program, and therefore something that should be dealt with subsequently. But the attributes of the range in question are unknown as a result in any case, and so whether the page table walker's view of this limbo state matches the CPU's is irrelevant. Ergo, the barrier can be skipped in case of failure.